builq.go

```diff
@@ -35,86 +38,120 @@ type constString string
 // Addf formats according to a format specifier, writes to query and appends args.
 // Format param must be a constant string.
 func (b *Builder) Addf(format constString, args ...any) *Builder {
-	if b.err != nil {
-		return b
-	}
-	// prealloc on first Addf
-	if b.args == nil {
-		b.query.Grow(100)
-		b.args = make([]any, 0, 10)
-	}
-
-	wargs := make([]any, len(args))
-	for i := range args {
-		wargs[i] = &argument{value: args[i], builder: b}
+	if b.formats == nil {
+		b.formats = make([]string, 0, 10)
+		b.rawArgs = make([][]any, 0, 10)
 	}
+	b.formats = append(b.formats, string(format))
+	b.rawArgs = append(b.rawArgs, args)
+	return b
+}
 
-	_, err := fmt.Fprintf(&b.query, string(format), wargs...)
-	if b.err == nil && err != nil {
-		b.err = err
-	}
+func (b *Builder) Build() (string, []any, error) {
+	query := b.build()
+	return query, b.args, b.err
+}
 
-	b.query.WriteByte('\n')
-	return b
+func (b *Builder) DebugBuild() string {
+	b.debug = true
+	query := b.build()
+	b.debug = false
+	return query
 }
 
-var missingRE = regexp.MustCompile(`%!.\(MISSING\)`)
+func (b *Builder) build() string {
+	sb := make([]string, 0, len(b.formats))
 
-func (b *Builder) Build() (string, []any, error) {
-	// prioritize the errors from the query string,
-	// otherwise Build might return a wrong error.
-	query := b.query.String()
-
-	switch {
-	case missingRE.MatchString(query):
-		b.err = errTooFewArguments
-	case strings.Contains(query, "%!(EXTRA"):
-		b.err = errTooManyArguments
-		// TODO(junk1tm): investigate panic cases
-		// case strings.Contains(query, "(PANIC="):
-		// 	_, msg, _ := strings.Cut(query, "(PANIC=")
-		// 	b.err = errors.New(msg)
+	for i := range b.formats {
+		format := b.formats[i]
+		args := b.rawArgs[i]
+
+		argID := -1
+		res := re.ReplaceAllStringFunc(format, func(s string) string {
+			argID++
+
+			if len(args) == 0 || len(args) == argID {
+				b.setErr(errTooFewArguments)
+				return ""
+			}
+			return b.write(s, args[argID])
+		})
+
+		if len(args) > 0 && argID != len(args)-1 {
+			b.setErr(errTooManyArguments)
```

Comment:
Passing an error through `Builder` was a hack due to `fmt.Formatter` being unable to return it, now we could return it in the last argument, as usual, WDYT?